File: src/lexer.rb
module TeaScript
  class Lexer
    attr_accessor :input, :index, :line, :column, :tokens, :state

    def self.re(ary, bound = nil)
      Regexp.new('\A(' + ary.map { |t| Regexp.escape(t) }.join('|') + ')' + (bound ? '\b' : ''))
    end

    KEYWORDS_RE  = re(KEYWORDS, true)
    OPERATORS_RE = re(OPERATORS)
    PARENS_RE    = re(PARENS)

    REGEXP = %w{
      ( , = : [ ! & | ? { } ; ~ + - * % ^ < >
      return when if else elsif unless
    }

    def initialize(input)
      self.input  = input.chomp
      self.index  = 0
      self.line   = 1
      self.column = 1
      self.tokens = []
      self.state  = []
    end

    def tokenize
      loop do
        str = input[self.index..-1]

        if str.nil? or str.empty?
          tokens << token(:EOF)
          break
        end

        case state.last
        when :embed
          next if match_embed(str) == :skip
        end

        match(str)
      end

      return Rewriter.new(tokens).rewrite
    end

    def match(str)
      tok = case str
      when /\A((?:#.*\n\s*)*#[^\n]*)/    then :COMMENT
      when /\A\n/                        then linefeed
      when /\A(;)/                       then $1
      when /\A(true|false)\b/            then :BOOL
      when /\A(null|nil|undefined)\b/    then :NIL
      when /\A(\d+)\b/                   then :NUMBER
      when /\A((\d+)?\.\d+)\b/           then :NUMBER
      when /\A\//                        then regexp(str); return
      when KEYWORDS_RE                   then $1.to_sym
      when OPERATORS_RE                  then $1
      when PARENS_RE                     then $1
      when /\A([a-zA-Z][a-zA-Z0-9]*)\b/  then :identifier
      when /\A([ \t\r]+)/                then :whitespace
      #when /\A('(\.|[^\'])*')/           then :STRING
      when /\A(')/                       then string(str); return
      when /\A(")/                       then embedable_string(str); return
      else throw :unknown_token
      end
      tokens << token(tok, $1) if tok
    end

    def match_embed(str)
      case str
      when /\A#\{/
        tokens << token('+', '+')
        tokens << token('(', '(')
        :skip
      when /\A\}/
        if str[1] == '"'
          tokens << token(')', ')')
          tokens << token(')', ')')
          state.pop() # :embed
          state.pop() # :string
        else
          # tokens << token(')', ')')
          # tokens << token('+', '+')
          # tokens << token('"', '"')
          consume(1)
          unput(') + "')
          state.pop() # :embed
        end
        :skip
      end
    end

    def linefeed
      self.line += 1
      self.column = 0
      consume(1)
      :LF
    end

    def embedable_string(str)
      string(str)
      if state.last == :string
        tokens << token(')', ')')
        state.pop
      end
    end

    def string(str)
      quote, idx, value = str[0], 1, ""
      re = quote == '"' ? /\A([^"]*")/ : /\A([^']*')/

      loop do
        if str[idx..-1] =~ re
          text = $1

          if quote == '"' and (j = (text =~ /#\{/))
            value += text[0..(j - 1)] + quote
            embed
            break
          else
            value += text

            if text.end_with?("\\#{quote}")
              idx += text.size
            else
              break
            end
          end
        else
          throw :unterminated_string
        end
      end

      tokens << token(:STRING, quote + value)
    end

    def embed
      unless state.last == :string
        tokens << token('(', '(')
        state << :string
      end
      state << :embed
      consume(-2)
    end

    def regexp(str)
      if !tokens.last or REGEXP.include?(tokens.last.name)
        if str =~ /\A(\/(\.|[^\\\/])*\/[gimy]*)/
          tokens << token(:REGEXP, $1)
          return
        end
      end
      tokens << token('/', '/')
    end

    def token(tok, value = nil)
      consume(value.size) if value
      Token.new(tok, value , line, column)
    end

    def consume(size)
      self.index  += size
      self.column += size
    end

    def unput(str)
      input.insert(self.index, str)
    end
  end
end

File: lib/lexer.js
var RE, REGEXP, Token, rewrite, words;
words = require('./lexer/words');
rewrite = require('./lexer/rewrite');
Token = require('./lexer/token');
RE = require('./lexer/definition').RE;
REGEXP = require('./lexer/definition').REGEXP;

function Lexer(input) {
    var self = this;
    self.input = input;
    self.index = 0;
    self.line = 1;
    self.column = 1;
    self.tokens = [];
    self.state = [];
    self.state.is = function (name) {
        return this[this.length - 1] === name;
    };
}
Lexer.prototype.constructor = Lexer;
Lexer.prototype.tokenize = function () {
    var self = this;
    var str;
    while (true) {
        str = self.input.slice(self.index);
        if (!(str)) {
            self.tokens.push(self.token('EOF'));
            break;
        }
        if (self.state.is('embed') && self.match_embed(str)) {
            continue;
        }
        self.match(str);
    }
    return rewrite(self.tokens);
};
Lexer.prototype.match = function (str) {
    var self = this;
    var m, name;
    if (m = str.match(RE.COMMENT)) {
        name = 'COMMENT';
    } else if (str.match(RE.LINEFEED)) {
        self.linefeed();
        return;
    } else if (m = str.match(RE.SEMICOLON)) {
        name = m[0];
    } else if (m = str.match(RE.BOOL)) {
        name = 'BOOLEAN';
    } else if (m = str.match(RE.NIL)) {
        name = 'NIL';
    } else if (m = str.match(RE.FLOAT)) {
        name = 'NUMBER';
    } else if (m = str.match(RE.INTEGER)) {
        name = 'NUMBER';
    } else if (str.match(RE.REGEXP)) {
        self.regexp(str);
        return;
    } else if (m = str.match(RE.KEYWORDS)) {
        name = m[0];
    } else if (m = str.match(RE.OPERATORS)) {
        name = m[0];
    } else if (m = str.match(RE.PARENS)) {
        name = m[0];
    } else if (m = str.match(RE.IDENTIFIER)) {
        name = 'identifier';
    } else if (m = str.match(RE.WHITESPACE)) {
        name = 'whitespace';
    } else if (str.match(RE.STRING)) {
        self.string(str);
        return;
    } else if (str.match(RE.STRING2)) {
        self.embedable_string(str);
        return;
    } else {
        m = str.match(/^(.+)\b/);
        throw Error(("unknown token " + (m[0]) + " at " + (self.line) + "," + (self.column)));
    }
    if (name) {
        self.tokens.push(self.token(name, m[0]));
    }
};
Lexer.prototype.match_embed = function (str) {
    var self = this;
    if (str.match(/^#\{/)) {
        self.tokens.push(self.token('+', '+'));
        self.tokens.push(self.token('(', '('));
    } else if (str.match(/^\}/)) {
        if (str[1] === '"') {
            self.tokens.push(self.token(')', ')'));
            self.tokens.push(self.token(')', ')'));
            self.state.pop();
            self.state.pop();
        } else {
            self.tokens.push(self.token(')', ')'));
            self.tokens.push(self.token('+', '+'));
            self.consume(-1);
            self.unput('"');
            self.state.pop();
        }
    } else {
        return false;
    }
    return true;
};
Lexer.prototype.linefeed = function () {
    var self = this;
    self.line += 1;
    self.column = 0;
    self.consume(1);
    self.tokens.push(self.token('LF'));
};
Lexer.prototype.embedable_string = function (str) {
    var self = this;
    self.string(str);
    if (self.state.is('string')) {
        self.tokens.push(self.token(')', ')'));
        self.state.pop();
    }
};
Lexer.prototype.string = function (str) {
    var self = this;
    var idx, j, m, quote, re, text, value;
    quote = str[0];
    idx = 1;
    value = '';
    re = quote === '"' ? /^([^"]*")/ : /^([^']*')/;
    while (true) {
        if (m = str.slice(idx).match(re)) {
            text = m[0];
            if (quote === '"' && (j = text.match(/#\{/))) {
                value += text.slice(0, j.index) + quote;
                self.embed();
                break;
            } else {
                value += text;
                if (text.slice(-2) === ("\\" + (quote))) {
                    idx += text.length;
                } else {
                    break;
                }
            }
        } else {
            throw Error("unterminated string");
        }
    }
    self.tokens.push(self.token('STRING', quote + value));
    self.line += (value.match(/\n/g) || '').length;
    if (self.state.is('string')) {
        self.tokens.push(self.token(')', ')'));
        self.consume(-1);
        self.state.pop();
    }
};
Lexer.prototype.embed = function () {
    var self = this;
    if (self.state.is('string')) {
        self.consume(-1);
    } else {
        self.tokens.push(self.token('(', '('));
        self.state.push('string');
        self.consume(-2);
    }
    self.state.push('embed');
};
Lexer.prototype.regexp = function (str) {
    var self = this;
    var i, idx, m, mm;
    i = self.tokens.length - 1;
    idx = 1;
    while (self.tokens[i] && self.tokens[i].name === 'whitespace') {
        i -= 1;
    }
    if (!self.tokens[i] || REGEXP.contains(self.tokens[i].name)) {
        while (true) {
            m = str.slice(idx).indexOf('/');
            if (m === -1) {
                throw Error('unterminated regexp');
            } else {
                idx += m + 1;
                if (str.slice(idx - 2, idx) !== '\\/') {
                    if (mm = str.slice(idx).match(/^[gimy]*/)) {
                        idx += mm[0].length;
                    }
                    self.tokens.push(self.token('REGEXP', str.slice(0, idx)));
                    return;
                }
            }
        }
    }
    self.tokens.push(self.token('/', '/'));
};
Lexer.prototype.token = function (name, value) {
    var self = this;
    if (value == null) value = null;
    if (value) {
        self.consume(value.length);
    }
    return new Token(name, value, self.line, self.column);
};
Lexer.prototype.consume = function (size) {
    var self = this;
    self.index += size;
    self.column += size;
};
Lexer.prototype.unput = function (str) {
    var self = this;
    if (self.index > 0) {
        self.input = self.input.slice(0, self.index) + str + self.input.slice(self.index);
    } else {
        self.input = str + self.input.slice(self.index);
    }
};
module.exports = {
    setInput: function (input) {
        this.tokens = new Lexer(input).tokenize();
    },
    lex: function () {
        var token;
        token = this.tokens.shift();
        if (token) {
            this.yytext = token.value;
            this.yyleng = token.value ? token.value.length : 0;
            this.yylineno = token.line - 1;
            this.yycolumn = token.column;
            return token.name;
        } else {
            return 'EOF';
        }
    }
};